PhysicsTools/TensorFlow/interface/TBBThreadPool.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * Custom TensorFlow thread pool implementation that schedules tasks in TBB.
+ * Based on TensorFlow 2.1.
+ * For more info, see https://gitlab.cern.ch/mrieger/CMSSW-DNN.
+ *
+ * Author: Marcel Rieger
+ */
+
+#ifndef PHYSICSTOOLS_TENSORFLOW_TBBTHREADPOOL_H
+#define PHYSICSTOOLS_TENSORFLOW_TBBTHREADPOOL_H
+
+#include "tensorflow/core/lib/core/threadpool.h"
+
+#include "tbb/task_scheduler_init.h"
+#include "tbb/task_arena.h"
+#include "tbb/task_group.h"
+
+namespace tensorflow {
+
+  class TBBThreadPool : public tensorflow::thread::ThreadPoolInterface {
+  public:
+    static TBBThreadPool& instance(int nThreads = -1) {
+      static TBBThreadPool pool(nThreads);
+      return pool;
+    }
+
+    explicit TBBThreadPool(int nThreads = -1) : nThreads_(nThreads), numScheduleCalled_(0) {
+      // when nThreads is zero or smaller, use the default value determined by tbb
+      if (nThreads_ <= 0) {
+        nThreads_ = tbb::task_scheduler_init::default_num_threads();
+      }
+    }
+
+    void Schedule(std::function<void()> fn) override {
+      numScheduleCalled_ += 1;
```

Comment:
Given that `TBBThreadPool` (and `NoThreadPool`) are singletons, should `Schedule()` be thread safe?`Or does TensorFlow itself guarantee that `Schedule()` will never be called concurrently even if we're running inference for many models and events concurrently?

Reply:
I am not entirely sure about this. Under the assumption that tbb doesn't use any internal singletons itself, the only non-local object to protect would be `numScheduleCalled_`, so we could make it atomic.

Reply:
> the only non-local object to protect would be `numScheduleCalled_`, so we could make it atomic.

That's exactly what I was thinking. Then also the function-local statics could be declared `CMS_THREAD_SAFE` to suppress warnings from the static analyzer.